pkg/provider/terraform/instance/plugin.go

```diff
@@ -1172,75 +1134,146 @@ func parseAttachTag(tf *TFormat) ([]string, error) {
 	return []string{}, nil
 }
 
+// External functions using during describe; broken out for testing
+type describeFns struct {
+	tfShow func(resTypes []TResourceType, propFilter []string) (map[TResourceType]map[TResourceName]TResourceProperties, error)
+}
+
 // DescribeInstances returns descriptions of all instances matching all of the provided tags.
 func (p *plugin) DescribeInstances(tags map[string]string, properties bool) ([]instance.Description, error) {
+	fns := describeFns{
+		tfShow: p.doTerraformShow,
+	}
+	return p.doDescribeInstances(fns, tags, properties)
+}
+
+// doDescribeInstances returns descriptions of all instances matching all of the provided tags.
+func (p *plugin) doDescribeInstances(fns describeFns, tags map[string]string, properties bool) ([]instance.Description, error) {
 	logger.Debug("DescribeInstances", "tags", tags, "V", debugV1)
-	// Acquire lock since we are reading all files and potentially running "terraform show"
-	p.fsLock.Lock()
-	defer p.fsLock.Unlock()
+	// Get the read lock and then check if the cache is populated
+	p.fsLock.RLock()
+
+	if p.cachedInstances == nil {
+		// Cache is not populated, now we need the write lock (after releasing the read lock) and
+		// then we can attempt to populate and use the cache
+		p.fsLock.RUnlock()
+		p.fsLock.Lock()
```

Comment:
This code here that tries to refresh the cache by first obtaining the lock can  be moved into `refreshCachedInstances()` so that the write lock is acquired and released on entry/exit of that function.   Then the code here can just work with the read lock/unlock to guard the critical section.  OR you could have a function `isCacheNil()` that inside that function it's doing the acquire/release of read lock.  

This way we also have corresponding acquire/release of the locks, regardless of the lock type.  

What do you think?

Reply:
The problem is that the lock is not re-entrant; so we cannot acquire the write lock if we already have the read lock. I wanted move the write lock acquisition into a function but I found myself always needed to call that function _after_ I had the read lock.

Reply:
I am thinking

```
func (p *plugin) isCacheNil() bool {
   p.lock.RLock()
   defer p.lock.Unlock()
   return p.cache == nil 
}

func (p *plugin) refreshCache() {
    p.lock.Lock() // write lock
    defer p.lock.Unlock()
    // check again after acquiring write lock
    if p.cache == nil {
        // do work
    }  
}

func (p *plugin) doDescribeInstances() {
   p.isCacheNil() { 
       p.refreshCache()        
   }
   // by now, we should have cached value in any case (nil or not nil)
   // there's no lock from this scope.

   p.lock.RLock()
   defer lock.RUnlock()
   // do filtering / computing a view of the cached value
}
```
 Will this work?

Reply:
@chungers Yes, thank you! See new commit that updated to this approach.